### DISCUSSION TITLE
Add password to Application

### DISCUSSION THREAD

**Discussion by collin-catalogic** [microsoftgraph/msgraph-sdk-go#87] [Q&A] [answered]
Using this SDK, i am trying to set password to an application following [docs](https://docs.microsoft.com/en-us/graph/api/application-addpassword?view=graph-rest-1.0&tabs=http) mentioned here:
```Golang
appPostOptions := &applications.ApplicationsRequestBuilderPostOptions{
	Body: graph.NewApplication(),
}
appPostOptions.Body.SetDisplayName(to.StringPtr(displayName))
azureApp, err := a.GraphServiceClient.Applications().Post(appPostOptions)

// Adding password to the created app
requestBody := addpassword.NewPasswordCredentialRequestBody()
passwordCredential := graph.NewPasswordCredential()
requestBody.SetPasswordCredential(passwordCredential)

password := uuid.New()
ptrPassword := to.StringPtr(password.String())
passwordCredential.SetKeyId(ptrPassword)
passwordCredential.SetSecretText(ptrPassword)

options := &addpassword.AddPasswordRequestBuilderPostOptions{
        Body: requestBody,
}

passwordCreds, err := a.GraphServiceClient.ApplicationsById(*azureApp.GetId()).AddPassword().Post(options)
```
Im getting this error: The server returned an unexpected status code and no error factory is registered for this code: 400

From the docs, ```requestBody := msgraphsdk.New()``` seems to be outdated (not sure)
Can i please get some help in this implementation.

**Comment by baywet** [comment]
Hi @collin-catalogic ,
Thanks for trying the Go SDK and for reaching out.
There are multiple problems this discussion outlines:

## Snippet in the documentation

The snippet generation (all code snippets are generated in the docs) doesn't seem to be adding the struct name for request bodies when calling the constructor. [I've logged an issue for that in the snippets generation repo](https://github.com/microsoftgraph/microsoft-graph-devx-api/issues/890)

## Error handling

We've recently [added error handling in kiota](https://github.com/microsoft/kiota/pull/1100), the generator for this SDK. However this feature requires an update to the OpenAPI description we're using to generate the SDK, the [change to the conversion library used to generate that description](https://github.com/microsoft/OpenAPI.NET.OData/pull/167) has been made, but since the conversion library [hasn't been released yet because it's pending on some more work](https://github.com/microsoft/OpenAPI.NET.OData/milestone/5), we're not seeing the full benefits.
What's happening right now is your request is failing for some reason, but because no error types are registered for that operation, the SDK client returns you a generic error instead of the detailed error information from the API. We're working to get this lib released so errors are more practical in the SDKs.

## API call

I'm not sure why the API call is failing itself, could be a permission issue, could be a payload issue, could be something else. The best avenue to look into that at the moment would be to use graph explorer or postman to issue the same exact request and get the actual error message returned by the API.

## Curiosity question

In `to.StringPtr()` what's the import behind `to`?

**Comment by collin-catalogic** [comment]
Sure, i will try with the graph explorer for the same API.
The import for ```to``` is ```"github.com/Azure/azure-sdk-for-go/sdk/to"```

**Comment by collin-catalogic** [comment, marked as answer]
I tried using the Graph Explorer, so, if secretText is set in the request body, it throws an error saying:
```
{
    "error": {
        "code": "Request_BadRequest",
        "message": "The password value is automatically generated and should not be specified.",
        "innerError": {
            "date": "2022-02-25T15:51:59",
            "request-id": "d2e47cef-edd0-4bed-85ab-921469c2cc37",
            "client-request-id": "2c3bfc3b-64ab-a9d4-9481-a9622492291d"
        }
    }
}
```
The request body that works is:
```
{
    "passwordCredential": {
        "displayName": "Password friendly name",
        "endDateTime": "2024-02-25T15:50:40.4213847Z",
        "startDateTime": "2022-02-25T15:50:40.4213847Z"
    }
}
```
Code snippet that worked for me:
```
	options := &addpassword.AddPasswordRequestBuilderPostOptions{
		Body: addpassword.NewPasswordCredentialRequestBody(),
	}
	passwordCredential := graph.NewPasswordCredential()
	displayName := "Password Credentials"
	passwordCredential.SetDisplayName(&displayName)
	options.Body.SetPasswordCredential(passwordCredential)
	passwordCreds, err := graphServiceClient.ApplicationsById(appID).AddPassword().Post(options)
```
But when i add these lines:
```
	startDateTime := time.Now()
	endDateTime := time.Now().Add(time.Minute * 710)
	passwordCredential.SetStartDateTime(&startDateTime)
	passwordCredential.SetEndDateTime(&endDateTime)
```
It throws error "400"

**Reply by baywet** [reply]
ok that makes sense the API doesn't want your application to set the password as part of its functional requirements. Nothing wrong with the SDKs besides the missing specialized error types we're already working on.

On the dates aspect it could be either of two things:
- the API has other functional requirements on the date ranges (too short/too long etc...)
- serialization is not doing the right thing.

Could you use the JSON Serialization writer (github.com/microsoft/kiota/serialization/go/json) to write out (write object value) the request body and log it? and share the result here please?

**Reply by collin-catalogic** [reply]
I could not figure out how to print the request body, help please.

**Reply by baywet** [reply]
it should be something like

```Golang
writer := json.NewJsonSerializationWriter()
err := writer.WriteObjectValue(requestBody)
content, err := writer.GetSerializedContent()
strContent := string(content)
//log str content
```

where json is an alias import of `github.com/microsoft/kiota/serialization/go/json`
